why-pytest/index.qmd

```diff
@@ -0,0 +1,166 @@
+
+## Decision drivers
+
+::: content-hidden
+List some reasons for why we need to make this decision and what things
+have arisen that impact work.
+:::
+
+-   Knowing that software works as intended requires some form of
+    testing.
+-   Developing software is easier when we can compare the expected
+    output with the actual output in a formal testing approach.
+-   Identifying bugs is easier when there are tests that check the
+    expected functionality of our code.
+-   Identifying breaking changes is also easier when there are tests.
+
+## Considered options
+
+::: content-hidden
+List and describe some of the options, as well as some of the benefits
+and drawbacks for each option.
+:::
+
+Django's default unit testing framework is
+[unittest](https://docs.python.org/3/library/unittest.html), which is
+also Python's built-in testing framework. Aside from the default
+unittest, there is really only one other alternative for unit testing:
+
+-   [pytest](https://docs.pytest.org/en/8.2.x/) with
+    [django-pytest](https://pytest-django.readthedocs.io/en/latest/)
+
+### unittest
+
+[unittest](https://docs.python.org/3/library/unittest.html) is the
+built-in framework in Python. It was inspired by JUnit (the unit testing
+framework in Java) and has a similar structure to it, for better or
+worse.
+
+::::: columns
+::: column
+#### Benefits
+
+-   Comes by default with Django and Python.
+-   Documentation on the official Django website uses unittest.
+:::
+
+::: column
+#### Drawbacks
+
+-   Requires more extra code to setup and structure tests. For instance,
+    all tests need to be a `class` object and some testing needs
+    `setUp()` and `tearDown()` methods.
+-   The test output is not very readable and beginner friendly.
+-   Creating tests takes more effort and time.
+-   Writing tests follows more of the object-oriented programming style,
+    which can be harder to reason about and design tests for.
+:::
+:::::
+
+### pytest
+
+[pytest](https://docs.pytest.org/en/8.2.x/), along with the [Django
+plugin](https://pytest-django.readthedocs.io/en/latest/), is a newer
+testing framework that was built to improve on the limitations of the
+default unittest framework.
+
+::::: columns
+::: column
+#### Benefits
+
+-   Integrates with [unittest style
+    tests](https://docs.pytest.org/en/7.1.x/how-to/unittest.html), so
+    don't need to refactor all existing tests.
+-   Has easier to read output of the test results.
+-   Uses a simpler syntax for writing tests, with a simple `assert`
+    statement.
+-   Created and developed more recently and from scratch, so is better
+    designed than unittest.
+-   While it isn't the default unit testing framework in Python, it is
+    widely used.
+:::
+
+::: column
+#### Drawbacks
+
+-   Requires installing two additional packages: `pytest` and
+    `pytest-django`.
+-   Tests written in the pytest style can't be used by other non-unit
+    testing frameworks.
```

Comment:
I guess it's not 100% obvious to me what this means 😅 . I would understand "other unit testing frameworks"